sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -620,6 +632,93 @@ def _next_id(self):
     return str(self._last_id)
 
 
+class CachingMaterializingStateHandler(object):
+  """ A State handler which retrieves and caches state. """
+
+  def __init__(self, global_state_cache, underlying_state):
+    self._underlying = underlying_state
+    self._state_cache = global_state_cache
+    self._context = threading.local()
+
+  @contextlib.contextmanager
+  def process_instruction_id(self, bundle_id, cache_tokens):
+    if getattr(self._context, 'cache_token', None) is not None:
+      raise RuntimeError(
+          'Cache tokens already set to %s' % self._context.cache_token)
+    # TODO Also handle cache tokens for side input, if present:
+    # https://issues.apache.org/jira/browse/BEAM-8298
+    user_state_cache_token = None
+    for cache_token_struct in cache_tokens:
+      if cache_token_struct.HasField("user_state"):
+        # There should only be one user state token present
+        assert not user_state_cache_token
+        user_state_cache_token = cache_token_struct.token
+    try:
+      self._context.cache_token = user_state_cache_token
+      with self._underlying.process_instruction_id(bundle_id):
+        yield
+    finally:
+      self._context.cache_token = None
+
+  def blocking_get(self, state_key, coder, is_cached=False):
+    if not self._should_be_cached(is_cached):
+      # no cache / tokens, can't do a lookup/store in the cache
+      return self._materialize_iter(state_key, coder)
+    # Cache lookup
+    cache_state_key = self._convert_to_cache_key(state_key)
+    cached_value = self._state_cache.get(cache_state_key,
+                                         self._context.cache_token)
+    if cached_value is None:
+      # Cache miss, need to retrieve from the Runner
+      materialized = cached_value = list(
```

Comment:
This will break if the value is too large to fit into memory. Of particular concern is the case of a GBK result that doesn't fit into memory and is fetched via the state API, but very large bag states are not that uncommon either.

Reply:
Without a size optimization, I don't think there is much we can do at the moment. For large state, the continuation token could be used, but it is not utilized by the Runner for now.

Considering that state caching is disabled for now and the previous state materialization strategy (lazily fetch one item at a time) has not been altered, it seems fair to address this in a follow-up and mark this a TODO for now.

Reply:
Also reflected this in the comment here with a JIRA link. Just to restate, the materialization logic with disabled caching (default) is unaltered here.